File: src/core/git.ts
import { spawn } from "node:child_process";
import { createHash } from "node:crypto";
import { readFile } from "node:fs/promises";
import { join } from "node:path";
import { DetDocError } from "./errors.js";

export interface DirtyFile {
  path: string;
  status: string;
}

export class GitRepository {
  constructor(readonly cwd: string) {}

  async git(args: string[], options: { input?: string } = {}): Promise<string> {
    return new Promise((resolve, reject) => {
      const child = spawn("git", ["-c", "core.quotepath=false", ...args], {
        cwd: this.cwd,
        stdio: [options.input === undefined ? "ignore" : "pipe", "pipe", "pipe"],
      });
      const stdout: Buffer[] = [];
      const stderr: Buffer[] = [];

      child.stdout?.on("data", (chunk: Buffer) => stdout.push(chunk));
      child.stderr?.on("data", (chunk: Buffer) => stderr.push(chunk));
      child.on("error", (error) => reject(new DetDocError(`git ${args.join(" ")} failed: ${error.message}`, "GIT_FAILED")));
      child.on("close", (code) => {
        const out = Buffer.concat(stdout).toString("utf8");
        const err = Buffer.concat(stderr).toString("utf8");
        if (code === 0) {
          resolve(out);
        } else {
          reject(new DetDocError(`git ${args.join(" ")} failed with exit code ${code}: ${err}`, "GIT_FAILED"));
        }
      });

      if (options.input !== undefined) {
        child.stdin?.end(options.input);
      }
    });
  }

  async root(): Promise<string> {
    return (await this.git(["rev-parse", "--show-toplevel"])).trim();
  }

  async headCommit(): Promise<string> {
    return (await this.git(["rev-parse", "HEAD"])).trim();
  }

  async statusPorcelain(): Promise<DirtyFile[]> {
    const output = await this.git(["status", "--porcelain=v1"]);
    return output
      .split("\n")
      .filter(Boolean)
      .map((line) => ({ status: line.slice(0, 2), path: line.slice(3) }));
  }

  async diff(): Promise<string> {
    return this.git(["diff", "--no-color", "--no-ext-diff", "--binary", "--", "."]);
  }

  async diffNameOnly(): Promise<string[]> {
    const output = await this.git(["diff", "--name-only", "--", "."]);
    return output.split("\n").filter(Boolean);
  }

  async applyPatch(patch: string): Promise<void> {
    await this.git(["apply", "--whitespace=nowarn"], { input: patch });
  }

  async changedFilesFromPatch(patch: string): Promise<string[]> {
    const output = await this.git(["apply", "--numstat", "-"], { input: patch });
    return output
      .split("\n")
      .filter(Boolean)
      .map((line) => line.split("\t").at(-1))
      .filter((path): path is string => Boolean(path));
  }

  async fileSha256(relativePath: string): Promise<string | null> {
    try {
      const bytes = await readFile(join(this.cwd, relativePath));
      return createHash("sha256").update(bytes).digest("hex");
    } catch {
      return null;
    }
  }
}

File: src-tauri/src/detdoc/git.rs
use std::path::{Path, PathBuf};
use std::process::Command;

use super::error::{DetDocError, DetDocResult};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct GitStatusEntry {
    pub status: String,
    pub path: String,
}

#[derive(Debug, Clone)]
pub struct GitRepository {
    pub cwd: PathBuf,
}

impl GitRepository {
    pub fn new(path: impl AsRef<Path>) -> Self {
        Self { cwd: path.as_ref().to_path_buf() }
    }

    pub fn git(&self, args: &[&str]) -> DetDocResult<String> {
        let output = Command::new("git")
            .args(args)
            .current_dir(&self.cwd)
            .output()
            .map_err(|error| DetDocError::new("GIT_SPAWN_FAILED", error.to_string()))?;
        if !output.status.success() {
            return Err(DetDocError::new(
                "GIT_COMMAND_FAILED",
                format!("git {:?}: {}", args, String::from_utf8_lossy(&output.stderr)),
            ));
        }
        Ok(String::from_utf8_lossy(&output.stdout).to_string())
    }

    pub fn head(&self) -> DetDocResult<String> {
        Ok(self.git(&["rev-parse", "HEAD"])?.trim().to_string())
    }

    pub fn status_porcelain(&self) -> DetDocResult<Vec<GitStatusEntry>> {
        let output = self.git(&["status", "--porcelain", "-uall"])?;
        Ok(output
            .lines()
            .filter_map(|line| {
                if line.len() < 4 { return None; }
                Some(GitStatusEntry {
                    status: line[0..2].trim().to_string(),
                    path: line[3..].trim().to_string(),
                })
            })
            .collect())
    }

    pub fn apply_patch(&self, patch: &str) -> DetDocResult<()> {
        let mut child = Command::new("git")
            .args(["apply", "--binary", "-"])
            .current_dir(&self.cwd)
            .stdin(std::process::Stdio::piped())
            .stderr(std::process::Stdio::piped())
            .spawn()
            .map_err(|error| DetDocError::new("GIT_APPLY_SPAWN_FAILED", error.to_string()))?;
        use std::io::Write;
        child.stdin.as_mut().unwrap().write_all(patch.as_bytes()).map_err(|error| DetDocError::new("GIT_APPLY_STDIN_FAILED", error.to_string()))?;
        let output = child.wait_with_output().map_err(|error| DetDocError::new("GIT_APPLY_WAIT_FAILED", error.to_string()))?;
        if !output.status.success() {
            return Err(DetDocError::new("GIT_APPLY_FAILED", String::from_utf8_lossy(&output.stderr).to_string()));
        }
        Ok(())
    }
}
